### PR TITLE
magento/magento2#23847: Command "setup:db-declaration:generate-patch" does not add revert() method to patch

### PR DESCRIPTION
**Pull request by atwixfirster**
### Fixed Issues (if relevant)

1. magento/magento2#23847: Command "setup:db-declaration:generate-patch" does not add revert() method to patch

### Manual testing scenarios (*)

Example with `Magento_Cms`:
```
bin/magento setup:db-declaration:generate-patch --revertable=true Magento_Cms TestPatch
```

**Actual result:** 
`app/code/Magento/Cms/Setup/Patch/Data/TestPatch.php` is created with content without `revert()` method:

```php
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

namespace Magento\Cms\Setup\Patch\Data;

use Magento\Framework\Setup\Patch\DataPatchInterface;
use Magento\Framework\Setup\Patch\SchemaPatchInterface;
use Magento\Framework\Setup\Patch\PatchRevertableInterface;
use Magento\Framework\Setup\ModuleDataSetupInterface;

/**
* Patch is mechanism, that allows to do atomic upgrade data changes
*/
class TestPatch implements
    DataPatchInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }
}
```

**Expected result:**
`app/code/Magento/Cms/Setup/Patch/Data/TestPatch.php` is created with `revert()` method in the content:

```php
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

namespace Magento\Cms\Setup\Patch\Data;

use Magento\Framework\Setup\Patch\DataPatchInterface;
use Magento\Framework\Setup\Patch\SchemaPatchInterface;
use Magento\Framework\Setup\Patch\PatchRevertableInterface;
use Magento\Framework\Setup\ModuleDataSetupInterface;

/**
* Patch is mechanism, that allows to do atomic upgrade data changes
*/
class TestPatch implements
    DataPatchInterface, PatchRevertableInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * @inheritdoc
     */
    public function revert()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/Developer/Console/Command/GeneratePatchCommand.php [hunk 1]
```diff
@@ -105,6 +95,7 @@ protected function execute(InputInterface $input, OutputInterface $output) : int
     {
         $moduleName = $input->getArgument(self::MODULE_NAME);
         $patchName = $input->getArgument(self::INPUT_KEY_PATCH_NAME);
+        $includeRevertMethod = ($input->getOption(self::INPUT_KEY_IS_REVERTABLE) == "true");
```
Comment (dmytro-ch) [inline app/code/Magento/Developer/Console/Command/GeneratePatchCommand.php]: Please use strict comparison operator `===`.
Please also use single-quotes `'true'` to keep the same style within the same class unless three is no specific need to use double-quotes.
I would also recommend avoiding the usage of redundant parenthesis.

As for the implementation itself, I think, it should be possible to use the following condition:
```
if ($input->getOption(self:: INPUT_KEY_IS_REVERTABLE)) {
...
}
```

Please check other examples of `\Symfony\Component\Console\Input\InputInterface::getOption` method usage.

File: app/code/Magento/Developer/Console/Command/patch_template.php.dist [hunk 2]
```diff
@@ -15,7 +15,7 @@ use Magento\Framework\Setup\ModuleDataSetupInterface;
 * Patch is mechanism, that allows to do atomic upgrade data changes
 */
 class %class% implements
-    %patchInterface%
+    %patchInterface%%patchRevertableInterface%
```
Comment (dmytro-ch) [inline app/code/Magento/Developer/Console/Command/patch_template.php.dist]: I guess the %patchRevertableInterface% should be on a new line.
Here's an example of what I mean:
```
/**
 * Wrapper for GraphQl NonNull
 */
class NonNull extends \GraphQL\Type\Definition\NonNull implements
    WrappedTypeInterface,
    InputTypeInterface,
    OutputTypeInterface
{
```

**Review by dmytro-ch — CHANGES_REQUESTED** [review]
@atwixfirster, nice catch!
Could you please check the minor CR notices?
Please also check the automated test failures.

Thank you!

File: app/code/Magento/Developer/Console/Command/patch_template.php.dist (hunk 2 above)
Comment (atwixfirster) [reply app/code/Magento/Developer/Console/Command/patch_template.php.dist]: > I guess the %patchRevertableInterface% should be on a new line.

Not really, @dmytro-ch  :)

`%patchRevertableInterface%` is the alias of `, PatchRevertableInterface`.

So, with the `--revertable=true` flag a result will be:

```php
class TestPatch implements
    DataPatchInterface, PatchRevertableInterface
{
```

Without it:

```php
class TestPatch implements
    DataPatchInterface
{
```

So, I would like to leave this part as it is.

Thank you!

File: app/code/Magento/Developer/Console/Command/GeneratePatchCommand.php (hunk 1 above)
Comment (atwixfirster) [reply app/code/Magento/Developer/Console/Command/GeneratePatchCommand.php]: **done**

**Comment by atwixfirster** [comment]
>  Could you please check the minor CR notices?

**fixed**

File: app/code/Magento/Developer/Console/Command/patch_template.php.dist (hunk 2 above)
Comment (dmytro-ch) [reply app/code/Magento/Developer/Console/Command/patch_template.php.dist]: I mean that the  `PatchRevertableInterface` should be added on a new line.
Please, check the example above, it shows how the interfaces should be arranged.

According to PSR-2 coding style guide:

> Lists of implements MAY be split across multiple lines, where each subsequent line is indented once. When doing so, the first item in the list MUST be on the next line, and there MUST be only one interface per line.

Comment (atwixfirster) [reply app/code/Magento/Developer/Console/Command/patch_template.php.dist]: > I mean that the PatchRevertableInterface should be added on a new line.

**done**

**Command 1:**

```
bin/magento setup:db-declaration:generate-patch --revertable=true Magento_Cms TestPatch
```

**Bash output 1:**

```
Patch app/code/Magento/Cms/Setup/Patch/Data/TestPatch.php has been successfully generated.
```

**Result 1:**

```
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

namespace Magento\Cms\Setup\Patch\Data;

use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\DataPatchInterface;
use Magento\Framework\Setup\Patch\PatchRevertableInterface;

/**
* Patch is mechanism, that allows to do atomic upgrade data changes
*/
class TestPatch implements
    DataPatchInterface,
    PatchRevertableInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * @inheritdoc
     */
    public function revert()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }
}
```

**Command 2:**

```
bin/magento setup:db-declaration:generate-patch --revertable=false Magento_Cms TestPatch
```

**Bash output 2:**

```
Patch app/code/Magento/Cms/Setup/Patch/Data/TestPatch.php has been successfully generated.
```

**Result 2:**

```
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

namespace Magento\Cms\Setup\Patch\Data;

use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\DataPatchInterface;

/**
* Patch is mechanism, that allows to do atomic upgrade data changes
*/
class TestPatch implements DataPatchInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }
}
```

**Command 3:**

```
bin/magento setup:db-declaration:generate-patch --revertable=true --type=schema Magento_Cms TestPatch
```

**Bash output 3:**

```
Patch app/code/Magento/Cms/Setup/Patch/Schema/TestPatch.php has been successfully generated.
```

**Result 3:**

```
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

namespace Magento\Cms\Setup\Patch\Schema;

use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\PatchRevertableInterface;
use Magento\Framework\Setup\Patch\SchemaPatchInterface;

/**
* Patch is mechanism, that allows to do atomic upgrade data changes
*/
class TestPatch implements
    SchemaPatchInterface,
    PatchRevertableInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * @inheritdoc
     */
    public function revert()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }
}
```

**Command 4:**

```
bin/magento setup:db-declaration:generate-patch --revertable=false --type=schema Magento_Cms TestPatch
```

**Bash output 4:**

```
Patch app/code/Magento/Cms/Setup/Patch/Schema/TestPatch.php has been successfully generated.
```

**Result 4:**

```
<?php
/**
 * Copyright © Magento, Inc. All rights reserved.
 * See COPYING.txt for license details.
 */

namespace Magento\Cms\Setup\Patch\Schema;

use Magento\Framework\Setup\ModuleDataSetupInterface;
use Magento\Framework\Setup\Patch\SchemaPatchInterface;

/**
* Patch is mechanism, that allows to do atomic upgrade data changes
*/
class TestPatch implements SchemaPatchInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }
}

```

**Comment by dmytro-ch** [comment]
@atwixfirster, thank you for the fixes!
Could you please squash all commits into a single one in order to clean up history?

**Comment by atwixfirster** [comment]
>  Could you please squash all commits into a single one in order to clean up history?

**done**

Thank you, @dmytro-ch 👍

**Review by dmytro-ch — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @dmytro-ch, thank you for the review.
[ENGCOM-5619](https://jira.corp.magento.com/browse/ENGCOM-5619) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
Hi @atwixfirster 

During testing we faced the issue

**_Problem:_** `revert()` method generate to patch even if option `--revertable=false`

**Steps to reproduce:**
Example with Magento_Cms

`bin/magento setup:db-declaration:generate-patch --revertable=false Magento_Cms TestPatch
`

**Actual Result:**

**class** TestPatch implements
    DataPatchInterface,
    PatchRevertableInterface
{
    /**
     * @var ModuleDataSetupInterface $moduleDataSetup
     */
    private $moduleDataSetup;

    /**
     * @param ModuleDataSetupInterface $moduleDataSetup
     */
    public function __construct(ModuleDataSetupInterface $moduleDataSetup)
    {
        $this->moduleDataSetup = $moduleDataSetup;
    }

    /**
     * Do Upgrade
     *
     * @return void
     */
    public function apply()
    {
    }

    /**
     * @inheritdoc
     */
    public function revert()
    {
    }

    /**
     * {@inheritdoc}
     */
    public function getAliases()
    {
        return [];
    }

    /**
     * {@inheritdoc}
     */
    public static function getDependencies()
    {
        return [

        ];
    }

@atwixfirster Could you take a look, please?

Thanks!

**Comment by atwixfirster** [comment]
@engcom-Alfa 

> Problem: revert() method generate to patch even if option --revertable=false

Let me please explain. 

If `--revertable` is present in the `bin/magento setup:db-declaration:generate-patch` command then `revert()` method will be added.

So, the key here is `--revertable`. It may have any value. For example,

`--revertable=hello`.

Please let me know you you have any additional questions here.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed 

**Before:**
before

**After:**
after

**Pull request merged**